### PR TITLE
Make translucentSuperType handle match types

### LINKED ISSUE #7894: Override check does not account for match type.
## minimized code

```Scala
scala> case class Box[T](t: T)
// defined case class Box

scala> type Boxed[T <: Tuple] <: Tuple = T match {
     |   case Unit => Unit
     |   case h *: t => Box[h] *: Boxed[t]
     | }

scala> trait Cmp[T <: Tuple] { def cmp(t: T, b: Boxed[T]): Boolean }
// defined trait Cmp

scala> object UnitCmp extends Cmp[Unit] {
     |   def cmp(t: Unit, b: Unit): Boolean = true
     | }
2 |  def cmp(t: Unit, b: Unit): Boolean = true
  |      ^
  |error overriding method cmp in trait Cmp of type (t: Unit, b: Boxed[Unit]): Boolean;
  |  method cmp of type (t: Unit, b: Unit): Boolean has incompatible type
1 |object UnitCmp extends Cmp[Unit] {
  |       ^
  |object creation impossible, since def cmp(t: Unit, b: Boxed[Unit]): Boolean is not defined
  |(The class implements a member with a different type: def cmp(t: Unit, b: Unit): Boolean)

scala> object UnitCmp extends Cmp[Unit] {
     |   def cmp(t: Unit, b: Boxed[Unit]): Boolean = true
     | }
2 |  def cmp(t: Unit, b: Boxed[Unit]): Boolean = true
  |      ^
  |error overriding method cmp in trait Cmp of type (t: Unit, b: Boxed[Unit]): Boolean;
  |  method cmp of type (t: Unit, b: Unit): Boolean has incompatible type
1 |object UnitCmp extends Cmp[Unit] {
  |       ^
  |object creation impossible, since def cmp(t: Unit, b: Boxed[Unit]): Boolean is not defined
  |(The class implements a member with a different type: def cmp(t: Unit, b: Unit): Boolean)
```

## expectation

I'm not sure if it is just the limitation of `Match Type`  or a bug.

### LINKED ISSUE #8666: Match types used in methods cause issues with inheritance
## Minimized code

```Scala
class Foo[A, B]()

type FooSnd[X] = X match
  case Foo[_, b] => b

trait Bar[A]:
  def bar(h: FooSnd[A]): Int

val foo: Bar[Foo[String, Int]] = new Bar[Foo[String, Int]]:
  def bar(h: FooSnd[Foo[String, Int]]) = h
```

but it works fine if `bar` is pulled top-level:

```diff
- val foo: Bar[Foo[String, Int]] = new Bar[Foo[String, Int]]:
-   def bar(h: FooSnd[Foo[String, Int]]) = h
+ def bar(h: FooSnd[Foo[String, Int]]) = h
```

Then it compiles as expected.

## Output

```scala
[error] -- Error: /Users/fixel/.src/fix-me-olivier.scala:19:6
[error] 19 |  def bar(h: FooSnd[Foo[String, Int]]) = h
[error]    |      ^
[error]    |error overriding method bar in trait Bar of type (h: FooSnd[Foo[String, Int]]): Int;
[error]    |  method bar of type (h: Int): Int has incompatible type
```

## Expectation
Not sure - the error message needs improvement.

### PR DESCRIPTION
**Pull request by odersky**
This fixes several erasure related things, such as signature, or erasure itself.

Fixes #8666
Fixes #7894

### CODE REVIEW

**Comment by odersky** [comment]
Note: This is technically a binary incompatible change since it causes bridge methods to be created where none were created before. On the other hand, I have a hard time thinking of an example that compiled before and now compiles with different binary signatures.

**Comment by sjrd** [comment]
> This is technically a binary incompatible change since it causes bridge methods to be created where none were created before.

If it's only bridges, and only *added*, then it's only *forward* binary incompatible; it's backward compatible.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -4011,7 +4011,7 @@ object Types {
       case tycon: TypeRef if tycon.symbol.isOpaqueAlias =>
         tycon.translucentSuperType.applyIfParameterized(args)
       case _ =>
-        superType
+        tryNormalize.orElse(superType)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala:4014]: Why do this in translucentSuperType and not superType? At the very least it means the documentation of the base translucentSuperType in https://github.com/lampepfl/dotty/blob/f618dd66a93a962c99c22eb586dbc93d7d410988/compiler/src/dotty/tools/dotc/core/Types.scala#L1891-L1892 is now incorrect.

**Review by smarter — CHANGES_REQUESTED** [review]
I played a bit with it and couldn't find a binary- or tasty-incompatible difference either (but I have to say that I don't really understand in what situations we do or don't normalize match types).

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala:4022]: I fixed the comment, which now also explains why we can't do it in `superType`.

**Review by smarter — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala:4022]: As far as I can see we did not normalize match types when computing the signatures and erasures of inherited members. That means we did not detect logical overrides, and that means we did not generate bridges where they would have been necessary. 

Match types that can be reduced directly in signatures were reduced. For instance in
```scala

  type M[X] = X match
    case Int => String
    case _ => X

  def test(x: M[Int]): Int = ???
```
the signature of `test` is `String => Int`.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 2]
```diff
@@ -1888,7 +1888,15 @@ object Types {
       case st => st
     }
 
-    /** Same as superType, except that opaque types are treated as transparent aliases */
+    /** Same as superType, except for two differences:
+     *   - opaque types are treated as transparent aliases
+     *   - applied type are matchtype-reduced if possible
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Types.scala:1893]:
```suggestion
     *   - applied types are matchtype-reduced if possible
```

**Comment by liufengyun** [comment]
This PR has a big impact on performance when compiling stdlib:

http://dotty-bench2.epfl.ch/

**Comment by odersky** [comment]
I tracked it down to match type simplification for tuples, and it seems in particular provablyDisjoint.

**Pull request merged**